sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -409,47 +409,53 @@ case class GpuProjectAstExec(
     projectList.collect { case ne: NamedExpression => ne.toAttribute }
   }
 
-  override def internalDoExecuteColumnar() : RDD[ColumnarBatch] = {
+  override def internalDoExecuteColumnar(): RDD[ColumnarBatch] = {
+    child.executeColumnar().mapPartitions(buildRetryableAstIterator)
+  }
+
+  def buildRetryableAstIterator(
+      input: Iterator[ColumnarBatch]): Iterator[ColumnarBatch] with AutoCloseable = {
     val numOutputRows = gpuLongMetric(NUM_OUTPUT_ROWS)
     val numOutputBatches = gpuLongMetric(NUM_OUTPUT_BATCHES)
     val opTime = gpuLongMetric(OP_TIME)
     val boundProjectList = GpuBindReferences.bindGpuReferences(projectList, child.output)
     val outputTypes = output.map(_.dataType).toArray
-    val rdd = child.executeColumnar()
-    rdd.mapPartitions { cbIter =>
-      new Iterator[ColumnarBatch] with AutoCloseable {
-        private[this] var compiledAstExprs =
-          withResource(new NvtxWithMetrics("Compile ASTs", NvtxColor.ORANGE, opTime)) { _ =>
-            boundProjectList.safeMap { expr =>
-              // Use intmax for the left table column count since there's only one input table here.
-              expr.convertToAst(Int.MaxValue).compile()
-            }
+    new Iterator[ColumnarBatch] with AutoCloseable {
```

Comment:
Nit: This could leverage GpuColumnarBatchIterator to reduce some of the replicated boilerplate code.

Reply:
Updated